sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala

```diff
@@ -73,18 +73,25 @@ object EliminateView extends Rule[LogicalPlan] with CastSupport {
       }
       // Map the attributes in the query output to the attributes in the view output by index.
       val newOutput = output.zip(queryOutput).map {
-        case (attr, originAttr) if !attr.semanticEquals(originAttr) =>
+        case (attr, originAttr)
+          if attr.name != originAttr.name ||
+            attr.nullable != originAttr.nullable ||
+            attr.dataType != originAttr.dataType =>
           // `CheckAnalysis` already guarantees that the cast is a up-cast for sure.
           Alias(cast(originAttr, attr.dataType), attr.name)(exprId = attr.exprId,
             qualifier = attr.qualifier, explicitMetadata = Some(attr.metadata))
         case (_, originAttr) => originAttr
       }
-      Project(newOutput, child)
+      if (newOutput != queryOutput) {
+        Project(newOutput, child) -> output.zip(queryOutput)
```

Comment:
This is still incorrect as not all columns are being changed from `output` to `queryOutput`. How about
```
val newNode = if .... else ...
newNode -> output.zip(newNode.output)
```